### PR TITLE
Auto detect hardware acceleration

### PR DESCRIPTION
**Pull request by lukasf**
This is an initial version for HW detection.

- Detect available hardware acceleration formats
- Detect available resolution
- Check profiles where appropriate
- Check level for H264 and HEVC
- Check installation status of MPEG-2 and VP9 codec
- Uses delay-initialization and then caches values

### CODE REVIEW

**Comment by mcosmin222** [comment]
Looks cool!
I think we should expose the hardware availability for passthroughs to the public, so apps can display this in an UI if necessary. But this is more of a "nice to have" rather than something really important.

**Comment by lukasf** [comment]
First I wanted to make it public. But when it is public, it becomes part of the API. And I am not sure if this will be final. It might be that we later find out that we have to do things differently.

Most importantly, I have combined all the available HW profiles for one format into one "status" instance. I query the resolution for the first profile I get, and I assume the same resolution is valid for all other profiles of that format as well. I think this is a valid assumption. But it could turn out that there are HW decoders that can do e.g. 4K for VP9 8bit profile but only FHD for VP9 10bit profile. We might only find out when the PR is already merged. That is why I am hesitating to make this a public API.

**Comment by mcosmin222** [comment]
Makes sense. I will look into creating a series of test files. Perhaps we can use ffmpeg.exe for that.

**Comment by lukasf** [comment]
That would be great. Having a set of test files with different formats, profiles and resolutions would allow us to test all corner cases.

**Comment by mcosmin222** [comment]
Seems ffmpeg has some test files. I will try to get them (fate samples).

**Comment by mcosmin222** [comment]
Sorry, haven't had time to look into this. I will give it more time next week.

**Comment by mcosmin222** [comment]
Anyway so after messing around a bit, found this website

https://samples.ffmpeg.org/

unfortunately, it cannot be mirrored easily. But on the other hand, I don't think we do need to mirror it. But it is a good way to find files and test them.

**Comment by lukasf** [comment]
Other sample sources I frequently used:

https://samples.mplayerhq.hu/
https://kodi.wiki/view/Samples

**Comment by lukasf** [comment]
I now added max profile and level to config class for both H264 and HEVC. This will allow more fine granular control in case of manual config. Max level is also used for auto detection, since we cannot detect level support.

**Comment by mcosmin222** [comment]
Cool!
I haven't had time to do much testing for this. I am busy tracking down a weird bug...

**Comment by lukasf** [comment]
I found out that indeed there can be different max resolutions, depending on profile. E.g. on one PC, HEVC Main is possible in 8K, but HEVC Main10 is only possible in 4k. The latest changeset addresses this.

**Comment by mcosmin222** [comment]
Looks good so far.

**Comment by mcosmin222** [comment]
Hi @lukasf ,
Do you think this is ready for release? I could disseminate this to a small portion of my users for more tests.

**Comment by lukasf** [comment]
Actually, there is something holding me back here. Some PCs, especially laptops, have two GPUs. On later Windows 10 versions, it is possible to configure per-app, which GPU is used. In my code, I currently just get the "Default" device. It might be that the "Default" will already be what has been configured for the app, but I am not sure. It could also be that "Default" will always return the iGPU, even if the app is configured to use the dedicated GPU, then my solution would report wrong capabilities.

I still need to test this. If it turns out that "Default" does not work in all cases, then the solution would need quite some rework.

**Comment by mcosmin222** [comment]
I could test that on my laptop.

**Comment by mcosmin222** [comment]
I think the code works just fine.

But if you want to be sure, you can get the preferred device like so:

               

```
                        IDXGIAdapter * pAdapter;
			IDXGIFactory * pFactory = NULL;

			CreateDXGIFactory1(__uuidof(IDXGIFactory), (void**)&pFactory);
			
			if(pFactory->EnumAdapters(0, &pAdapter) != DXGI_ERROR_NOT_FOUND)
			{
				
				auto adapterDescription = DXGI_ADAPTER_DESC();
				pAdapter->GetDesc(&adapterDescription);
				
				
			}

```

With this particular thing, I realized the nvidia GPU is really  bad for video decoding.

**Comment by lukasf** [comment]
I also tested on my laptop and indeed the default device is what has been configured for the app. So for that scenario, the code works fine.

> What happens if the user has 2 different GPUs connected to 2 different monitors?

I really don't know and I don't even have a "normal" PC to test that scenario.

> What happens if the preferred GPU is changed during play?

I am pretty sure that these changes are only reflected on next start of the app.

> What happens when external GPUs get involved? What if the GPU gets detached?

Windows 1809 introduced new classes which allow us to register for display adapter changes, I already looked at those a while back. We could register for that event and then reset our cached HW acceleration status, so next video file would have the right values. But I am not sure if we should switch to that target version, only for this event. It is a pretty rare situation, which we cannot handle 100% cleanly anyways (e.g. if playback is already running, results are unpredictable).

**Comment by mcosmin222** [comment]
For the 1809 events, this could be handled by the consumer app. We could expose a reset hardware cache method and let the apps handle it themselves. It might be needed to recreate the media stream source anyway, and we cannot do that at our level.

**Comment by lukasf** [comment]
Good idea. There is already a function RefreshAsync() which will wipe the cache and re-initialize.

**Comment by mcosmin222** [comment]
I have discovered some pretty interesting things about multi GPU setups.
Apparently, nvidia GPUs are really terrible at doing video decoding. On both 960m and RTX2070, video is sometimes choppy, and playback is frozen when speed is above 1x. It unfreezes with a seek.

On the other hand, the intel GPU is pretty decent at doing this, and supports multiple profiles as well.

**Comment by mcosmin222** [comment]
@lukasf ,

Now that I've begun trying to add this to my app, I think it is best to expose the HardwareAccelerationStatus class to the public, as well as its instances created from CodecChecker (which are currently internal). The reason I think of it this way is that it can be displayed to advanced users, and will also help debugging.

This can be done with minimum modifications.

**Comment by mcosmin222** [comment]
I will also add a sample to listen for display changes...

**Comment by lukasf** [comment]
I have kept the HardwareAccelerationStatus classes internal because these classes are actually used internally and they could completely change if things are further evolved. Each of these classes currently aggregates one or more HW decoder interfaces, which seems to work well, but might not be in the long run. Also, some things are not easy to interpret. Profiles for example are stored as "int", but you'd need a string to visualize this in a GUI.

It is unfortunate that there is no "ExperimentalAttribute" which you could place on types or methods (similar to "Deprecated"), to get a compiler warning shown. Then devs would know that this could change at any time, but are still free to use it if they like.

**Comment by mcosmin222** [comment]
I see your point. I guess we can use a wrapper class HardwareAccelerationStatus, similar to what we did for charset encoding. This way HardwareAccelerationStatus remains internal, and we can expose additional metadata about the hardware support using the wrapper class. What do you think?

**Comment by mcosmin222** [comment]
Well, it was all fun and giggles until I realized that ffmpeg profiles all start from 0, they are different enum for each codec, at which point I basically gave up.

**Comment by mcosmin222** [comment]
@lukasf 
Check out branch https://github.com/ffmpeginteropx/FFmpegInteropX/tree/hardware-check-UI

**Comment by mcosmin222** [comment]
Added UI sample.

**Comment by mcosmin222** [comment]
OK, maybe that was not such a great idea. But perhaps we should add a flag in the VideoStreamInfo class to show if it is hardware accelerated or not. This will be not be dependent on our detection code.

What do you think @lukasf ?

**Comment by lukasf** [comment]
Your branch still basically exposed the internal structure to outside. There can be multiple decoders for one format, and I kind of group them by codec and aggregate data. This seems to work fine, but it could turn out that a different structure is needed. Then we'd have to maintain two different models, one internal and one for public use, and people might ask "why is this file not HW accelerated although the info class says so".

Adding a flag to the StreamInfo class is a good idea! That way we can visualize if a file is HW accelerated or not, but without all the problems. The only thing is that actually we can't be 100% sure that HW acceleration is really used. In the end we just pass the data downstream and think/hope it will be accelerated, based on our checks. The decoder might still decide to go SW for some reason. But there is nothing we can do about that, without rolling our own custom HW acceleration. So I'd say, let's just add the flag and set it to true, if our checks say so.

I still think that we could at a later point expose the HW detection. It's just that I wouldn't do this with the first version. Let's first get this out and become a bit mature. When we feel that this is really stable and solid, we can at a later point make it a public API.

**Comment by mcosmin222** [comment]
Cool then. I will expose the puny bool sometime this week.

**Comment by mcosmin222** [comment]
I think the correct flag would if it is it compressed sample provider or not.

**Comment by lukasf** [comment]
Ok you still named it HardwareAccelerated, that's fine with me.

Oh man so many changes required just to add a single parameter. Constructor chaining is a real mess in C++, especially when you need to modify each constructor twice, once in the header and once in the implementation.

**Comment by mcosmin222** [comment]
The other option is to add a one time setter method to set the value. I could revert to that if you feel the changes are too widespread. 

Yes, i agree it is overkill to modify so many things just for a parameter, this is why I prefer to write the c++ classes inline in the header: much less overhead, and the compilation time penalty is negligible.

**Comment by lukasf** [comment]
No no you can just leave it as it is. It's just reminded me of how painful it is to add more constructor parameters, I have been going down that rocky road myself lately ;). I had been thinking if it could be a better approach to use properties to set everything and then later call initialize to bind it all together. Or something like that. But it does not have to be now.

I remember we had those flicker problems which went away after going through UI thread, so maybe that could help here as well. Let me know how it goes.

**Comment by mcosmin222** [comment]
We could in theory use some C++ magic, like friend classes accessing private fields and acting as some sort of builders. But this feels like cheating to me lol.

**Comment by lukasf** [comment]
One other workaround I thought of was using some container class (e.g. "StreamParameters"), which contains all the required things. Then we have just one single instance being passed down. When we need more parameters, we just add them to the container and it's all done. It's also a bit like cheating but without any real hacks ;)

**Comment by mcosmin222** [comment]
Well that is exactly what we did with the Configuration class. I'd say it was a success.

**Comment by lukasf** [comment]
Yes you are right. Maybe we could apply the same pattern to our SampleProviders...

**Comment by mcosmin222** [comment]
Indeed. But we should merge all PRs first, as it will cause a hell of merge conflicts :)

**Comment by lukasf** [comment]
The other two open PRs do not change any code files, so they could even be left open. But we should not start such a refactoring before finalizing and merging this PR here.

I could see us merging this soon, but I'd like to give this a thorough code review first. It's been a while since I wrote this and I want to make sure it still all makes sense to me :)

File: FFmpegInterop/CodecChecker.h [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+		static property HardwareAccelerationStatus^ HardwareAccelerationVP8 { HardwareAccelerationStatus^ get() { return hardwareAccelerationVP8; } }
+		static property HardwareAccelerationStatus^ HardwareAccelerationMPEG2 { HardwareAccelerationStatus^ get() { return hardwareAccelerationMPEG2; } }
+
+		static void PerformCheckHardwareAcceleration()
+		{
+			hardwareAccelerationH264->Reset();
+			hardwareAccelerationHEVC->Reset();
+			hardwareAccelerationWMV3->Reset();
+			hardwareAccelerationVC1->Reset();
+			hardwareAccelerationVP9->Reset();
```
Comment (lukasf) [inline FFmpegInterop/CodecChecker.h]: Reset of VP8 status missing

Comment (lukasf) [reply FFmpegInterop/CodecChecker.h]: Fixed.

**Comment by mcosmin222** [comment]
Sorry i got carried away by other things and didn't get to see the reviews. :(

I will give this a test drive on my various devices, see how it measures up. This might take a few days...

**Comment by lukasf** [comment]
I have added a new commit. I changed the bool property for auto detection into an enum, which also supports force software decode and force system decode. This allows to do something like this in MediaOpened:

	```
		auto session = Player->PlaybackSession;
		if (mss.VideoStream && session.NaturalVideoHeight == 0 && session.NaturalVideoWidth == 0)
			{
				if (mss.VideoStream.DecoderEngine == DecoderEngine.SystemDecoder && !currentFile.ForceSoftwareDecode)
				{
                                    ...Open video again with ForceFFmpegSoftwareDecoder!
```

This works pretty well for me. If hardware decoding does not work for some reason (decoder does not like that specific file), we fallback to SW decode. This is very fast and not noticeable to user.

**Comment by lukasf** [comment]
With this change, I think I am ready to merge.

**Comment by lukasf** [comment]
Wow, I found out something new today. I noticed that my dev machine suddenly refused playing HEVC with HW decoder. Before I reinstalled the OS, I could play HEVC on GPU without buying the HEVC extension. Do I really need to pay 1$ now for something that was free before?

A google search turned out that there is also a free version of that extension available, but it is hidden in the store, you won't find it through search. It is also from Microsoft and seems to be the exact same thing, it even has the same package family name. The only difference is that it is available as x64 only, and it's called "HEVC Video Extensions _from Device Manufacturer_" as opposed to the normal "HEVC Video Extensions". It gets pre-installed by manufacturers on devices with HW decoder, but it can also be installed on any x64 PC.

I now updated the codec checker to verify installation of the HEVC decoder, and added a method to open the store page. On x64 it will open the free version, on other platforms, it opens the paid version.

**Comment by lukasf** [comment]
I have added a new CodecRequired event, which will be raised if HW acceleration is technically available but the codec store extension is missing. This allows for a really cool app integration. The event is raised whenever a codec is required, and allows you to directly open the store page of that extension. Check C# sample for how this can be used.

File: FFmpegInterop/CodecChecker.h [hunk 2]
```diff
@@ -376,14 +400,36 @@ namespace FFmpegInterop
 				if (codecId == AV_CODEC_ID_MPEG2VIDEO)
 				{
 					result &= CheckIsMpeg2VideoExtensionInstalled();
+					if (!result && !hasAskedInstallMpeg2Extension)
+					{
+						RaiseCodecRequired(CodecRequiredReason::HardwareAcceleration, "MPEG2", "MPEG2 Video Extension", "9n95q1zzpmh4");
```
Comment (JunielKatarn) [inline FFmpegInterop/CodecChecker.h]: Maybe it would be better to inject these IDs as static configuration values instead of hard-coding them?

Comment (lukasf) [reply FFmpegInterop/CodecChecker.h]: There are only 4 video extensions available in the Store, and all of them are from Microsoft. As long as there is not a single 3rd party alternative in the Store, I do not see a need to make this configurable...

**Comment by mcosmin222** [comment]
Sorry, I've been on vacation :)

These changes look good.
